pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,814 @@
+
+from . import ExtensionArray, Categorical
+
+_VALID_CLOSED = set(['left', 'right', 'both', 'neither'])
+_interval_shared_docs = {}
+_shared_docs_kwargs = dict(
+    klass='IntervalArray',
+    name=''
+)
+
+
+_interval_shared_docs['class'] = """%(summary)s
+
+.. versionadded:: %(versionadded)s
+
+.. warning::
+
+   The indexing behaviors are provisional and may change in
+   a future version of pandas.
+
+Parameters
+----------
+data : array-like (1-dimensional)
+    Array-like containing Interval objects from which to build the
+    %(klass)s.
+closed : {'left', 'right', 'both', 'neither'}, default 'right'
+    Whether the intervals are closed on the left-side, right-side, both or
+    neither.
+%(name)s\
+copy : boolean, default False
+    Copy the meta-data.
+dtype : dtype or None, default None
+    If None, dtype will be inferred
+
+    .. versoinadded:: 0.23.0
+
+Attributes
+----------
+left
+right
+closed
+mid
+length
+values
+is_non_overlapping_monotonic
+
+Methods
+-------
+from_arrays
+from_tuples
+from_breaks
+%(extra_methods)s\
+
+%(examples)s\
+
+Notes
+------
+See the `user guide
+<http://pandas.pydata.org/pandas-docs/stable/advanced.html#intervalindex>`_
+for more.
+
+See Also
+--------
+Index : The base pandas Index type
+Interval : A bounded slice-like interval; the elements of an IntervalIndex
+interval_range : Function to create a fixed frequency IntervalIndex
+cut, qcut : Convert arrays of continuous data into Categoricals/Series of
+            Intervals
+"""
+
+
+@Appender(_interval_shared_docs['class'] % dict(
+    klass="IntervalArray",
+    summary="Pandas array for interval data that are closed on the same side",
+    versionadded="0.23.0",
+    name='', extra_methods='', examples='',
+))
+@add_metaclass(_WritableDoc)
+class IntervalArray(IntervalMixin, ExtensionArray):
+    dtype = IntervalDtype()
+    ndim = 1
+    can_hold_na = True
+    _na_value = _fill_value = np.nan
+
+    def __new__(cls, data, closed=None, dtype=None, copy=False,
+                fastpath=False, verify_integrity=True):
+
+        if fastpath:
+            return cls._simple_new(data.left, data.right, closed,
+                                   copy=copy, dtype=dtype,
+                                   verify_integrity=False)
+
+        if isinstance(data, ABCSeries) and is_interval_dtype(data):
+            data = data.values
+        if isinstance(data, (cls, ABCIntervalIndex)):
+            left = data.left
+            right = data.right
+            closed = data.closed
+        else:
+
+            # don't allow scalars
+            if is_scalar(data):
+                msg = ("{}(...) must be called with a collection of some kind,"
+                       " {} was passed")
+                raise TypeError(msg.format(cls.__name__, data))
+
+            data = maybe_convert_platform_interval(data)
+            left, right, infer_closed = intervals_to_interval_bounds(data)
+
+            if (com._all_not_none(closed, infer_closed) and
```

Comment:
don't you have a _check_closed function? maybe remove that from Interval (class) and make it a free function?